### ISSUE TITLE
[Library Stabilization] What should we do with the `fits_*` methods in BigInteger?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#17702]
The following methods on the `bigint` type seem out of place:
- `bigint.fits_ulong_p()`
- `bigint.fits_slong_p()`
- `bigint.fits_uint_p()`
- `bigint.fits_sint_p()`
- `bigint.fits_ushort_p()`
- `bigint.fits_sshort_p()`

I can think of a few options:
- We could move them into the GMP module along with the `mpzStruct` and potentially "limbs" methods (see #17694 and #17695)
- We could define Chapel type equivalents (e.g. `fitsInt`, `fitsUint`, `fitsInt64` or something)
- We could do a combination of the previous two bullets
- We could just remove them

If we keep them, I think they should return booleans like the functions in #17701

**Comment by bradcray** [comment]
Aren't these beneficial from the perspective of "I'd like to store this `bigint` into an `int` but only if it fits... let me check.  Like if I were writing a `.safeCast()` method, I'd use this to determine whether I was going to be successful or not?  That seems useful to me, and more related to their mathematical nature ("what is the value of this thing and does it fit?") than their implementation.  I'd preserve them as a result.

(having invoked `.safeCast()`, I think I might just provide the `.safeCast()` overloads that did this for the user as well as a convenience; but I guess that's not a stabilization question.  And I'd still provide these predicates even with the safeCasts).

**Comment by lydia-duncan** [comment]
@tiger-yash - I do think it would be appropriate to change the return type to bool.  The main questions to me are:
- Should these methods live in GMP instead, since they mostly apply to C types?
- Should there be versions of these methods that correspond to Chapel types?  (Those I would expect to live in the BigInteger module)

**Comment by bradcray** [comment]
At first glance, the first two bullets in the OP resonated with me.  Specifically:

* I don't think we should have Chapel routines that refer to the C types on bigints.  Putting them in the GMP module seems reasonable.
* I think that having Chapel equivalents of these routines seems convenient.

But then, thinking about the second one more, I also found myself wondering:

* if the definition is simply `myBigInt <= max(int) && myBigInt >= min(int)`  (or, replace `int` with any other integral type `t`), then maybe it's not so interesting?  (though still more convenient than typing that up).  I could imagine an argument for having these routines in C being that C's types are of less predictable size and the min/max values are a bit more awkward to get at.

If we created a Chapel equivalent, I'd probably make it something generic like `proc fitsIntegralType(type t: integral): bool` where the user would pass in the integral type they wanted to use.

**Comment by lydia-duncan** [comment]
> Keeping a chapel version of these methods does seem like a good choice as it is clean and simple. Also, leaving for GMP, do you mean by placing them with some other name or keeping the current versions as they are?

I realized I wasn't as clear in the previous comment as I should have been, sorry about that.  I've updated it slightly - my intention would be for there to be routines available in Chapel with this functionality.  I believe the GMP module already has `mpz_fits_ulong_p`, etc. in it.  Being as explicit as possible, we could:
1. consider that sufficient, asking the user to get the underlying `mpz` field from the `bigint` type in order to use them or
2. move the `bigint` methods that are currently defined (`bigint.fits_ulong_p`, etc.) into the GMP module and otherwise leave them untouched or
3. leave the methods in the BigInteger module, but change their names (in this route, I would expect the names to use camelCase but still refer to C types).

These options would not affect whether or not we add versions for Chapel types, I consider that a separate decision.

Personally, I favor option 2, but would be okay with option 1.  I would prefer not following option 3 and it sounds like Brad leans that way as well (though I'm not completely clear which of 1 and 2 he favors)

**Comment by lydia-duncan** [comment]
In the module re-review meeting, we decided to take the latest comment a step further - we'd like to make a single fits method that takes a type argument.  Options for the name:
- fits (2 votes)
- fitsType (3 votes)
- fitsInto (2 votes)

That seems a bit too close to call just yet

**Comment by lydia-duncan** [comment]
I find your justification for the name compelling :)

**Comment by bradcray** [comment]
I'm open to that method name proposal.

I balked slightly at:

> which would have overloads for each c-type.

since it seems like (a) we have this interface focus on Chapel types rather than C types (i.e., `int`, `int(8)`, etc. rather than `c_int`, c_short`, etc.) such that any conversion between Chapel and C types should happen under the covers and (b) rather than creating an overload per type, I'd probably just have a single routine that handled the different cases internally.  For example, possibly:

```chapel
proc bigint.fitsInto(type t: integral) {
  if t == int then ...
  else if t == int(32) then ...
}
```

Benefits being that the documentation will just show one routine rather than a routine per type.  I think by taking the formal take an integral type, we would also be enabling C integral types to be passed in if the user preferred (though I don't think we should call attention to that, just that it falls out).

**Comment by jeremiah-corrado** [comment]
I agree with (a) and (b). I was initially thinking of c_types because that's what the current interface is focused on; however, since we get those implicitly anyway as you pointed out, it definitely makes more sense to interface over integral types.

**Comment by bradcray** [comment]
> I was initially thinking of c_types because that's what the current interface is focused on; however, since we get those implicitly anyway as you pointed out, it definitely makes more sense to interface over integral types.

Generally speaking, I think we're trying to make most user-facing interfaces in Chapel standard modules use Chapel types unless those interfaces are themselves designed to be exposing C functionality.  So having the BigInteger API avoid C types seems like the ideal, using casts or safe casts when passing Chapel integer values downward to C APIs.  In contrast, since the GMP module is essentially making the C interface available to Chapel programmers, it is more reasonable/expected that it would use C types.

Combining that comment with some more thought about my sketch above, I think that the conditional within the routine should actually be something more like:

```chapel
if t == c_int then ...
else if t == c_long then ...
else ...
```

since the routines we'll be calling into will be expecting the C types.  And since the C type aliases have Chapel equivalents, this should still permit Chapel types to be passed in and compared.  So the main thing is not that we can't/shouldn't use C types in the implementation, just that we should try to avoid having the interface expose them (e.g., if we did `proc bigint.fitsInto(type t: c_int) { ... }`).

**Comment by lydia-duncan** [comment]
I think it'd be good to send an email out, not everyone follows every issue on Github.  But I suspect there won't be any objections

**Comment by lydia-duncan** [comment]
This was resolved by #20039
